### PR TITLE
Give a typical example of Makefile wrapper for coq_makefile (addresses #12903)

### LINKED ISSUE #12903: Why does coq_makefile generate `CoqMakeFile` rather than just `Makefile`?
It seems like it would be simpler just to generate `Makefile`.  Then you wouldn't need the extra arguments in "make **-f CoqMakefile**".  Does anyone know why it was done this way?

cc @herbelin

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** documentation

Fixes / closes #12903

- [X] Corresponding documentation was added / updated

### CODE REVIEW

File: doc/sphinx/practical-tools/utilities.rst [hunk 1]
```diff
@@ -92,14 +92,39 @@ CoqMakefile
...
+The recommended approach is to wrap ``CoqMakefile`` into a standard ``Makefile`` of the form:
+
+::
+
+    all: CoqMakefile
+            +make -f CoqMakefile all
+
+    clean: CoqMakefile
+            +make -f CoqMakefile clean
+            rm -f Makefile.coq
```
Comment (SkySkimmer) [inline doc/sphinx/practical-tools/utilities.rst:111]:
```suggestion
            rm -f CoqMakefile
```
?
Why CoqMakefile anyway, I've never seen it used unlike Makefile.coq?

File: doc/sphinx/practical-tools/utilities.rst [hunk 2]
```diff
@@ -92,14 +92,39 @@ CoqMakefile
...
+
+    CoqMakefile: _CoqProject
+            $(COQBIN)coq_makefile -f _CoqProject -o CoqMakefile
+
+    _CoqProject: ;
+
+    %: CoqMakefile
+            +make -f CoqMakefile $@
```
Comment (SkySkimmer) [inline doc/sphinx/practical-tools/utilities.rst:119]: Should we show a mechanism like https://github.com/coq/coq/blob/e583be62b74d71b5af159700e3a31f78fec9a7d2/Makefile.make#L166-L178 so that `make a b` results in a single subcall to make? Also
```suggestion
            $(MAKE) -f CoqMakefile $@
```
is the recommended way.

**Review by JasonGross — DISMISSED** [review]
Please don't duplicate the already-present example at https://coq.inria.fr/refman/practical-tools/utilities.html#reusing-extending-the-generated-makefile Instead link to it, and modify it if necessary.

**Comment by herbelin** [comment]
@JasonGross: Sorry, I did not see the example. So, I will simply link to it.

**Comment by herbelin** [comment]
And what about roughly inlining the section `Reusing/extending the generated Makefile` right after mentioning `CoqMakefile`, together with the existing example?

File: doc/sphinx/practical-tools/utilities.rst [hunk 3]
```diff
@@ -92,17 +92,55 @@ CoqMakefile
...
+      ####################################################################
+      ##                      Your targets here                         ##
+      ####################################################################
+
+      # This should be the last rule, to handle any targets not declared above
+      %: invoke-coqmakefile
+              @true
+
+An alternative would have been to give instead option ``-o Makefile``
+to ``coq_makefile`` but the advantage of a wrapper is that it provides
```
Comment (Zimmi48) [inline doc/sphinx/practical-tools/utilities.rst:136]: I understand everything starting with "the advantage of a wrapper" but nothing before. It would read fine if the sentence were just to start at this point.

Comment (herbelin) [reply doc/sphinx/practical-tools/utilities.rst:136]: OK, that's right that what comes before is in some sense implicit.

File: doc/sphinx/practical-tools/utilities.rst [hunk 4]
```diff
@@ -92,17 +92,55 @@ CoqMakefile
...
-  library is installed. Run ``make`` with the ``-f CoqMakefile``
-  option to use ``CoqMakefile``.
+  library is installed.
 
 CoqMakefile.conf
   contains make variables assignments that reflect
   the contents of the ``_CoqProject`` file as well as the path relevant to
   |Coq|.
 
+The recommended approach is to embed ``CoqMakefile`` into a standard
```
Comment (gares) [inline doc/sphinx/practical-tools/utilities.rst:102]:
```suggestion
The recommended approach is to invoke ``CoqMakefile`` from a standard
```

File: doc/sphinx/practical-tools/utilities.rst [hunk 5]
```diff
@@ -92,17 +92,55 @@ CoqMakefile
...
+
+An alternative would have been to give instead option ``-o Makefile``
+to ``coq_makefile`` but the advantage of a wrapper is that it provides
+a way independent of the version of Coq to regenerate a ``Makefile``
+specific to the current version of Coq. Additionally, the master
+``Makefile`` can be extended with targets not specific to Coq.
 
 An optional file ``CoqMakefile.local`` can be provided by the user in order to
-extend ``CoqMakefile``. In particular one can declare custom actions to be
+extend ``CoqMakefile`` with other Coq targets.
```
Comment (gares) [inline doc/sphinx/practical-tools/utilities.rst:142]: This is not correct. If you want to add a brand new target, you can just add it your wrapping makefile.
Makefile.local is to extend existing targets or set variables.

File: doc/sphinx/practical-tools/utilities.rst (hunk 4 above)
Comment (gares) [reply doc/sphinx/practical-tools/utilities.rst]: Also, I want to keep the sentence that says that including the generated makefile is discouraged.

File: doc/sphinx/practical-tools/utilities.rst [hunk 6]
```diff
@@ -453,50 +491,6 @@ line timing data:
       This target requires python to build the table.
 
 
-Reusing/extending the generated Makefile
-++++++++++++++++++++++++++++++++++++++++
-
-Including the generated makefile with an include directive is
```
Comment (gares) [inline doc/sphinx/practical-tools/utilities.rst:459]: I want to keep this sentence, please don't remove it.

File: doc/sphinx/practical-tools/utilities.rst (hunk 5 above)
Comment (herbelin) [reply doc/sphinx/practical-tools/utilities.rst:142]: I was rephrasing the text that followed: "Similarly one can customize the install target or even provide new targets." Does this sentence need to be changed also?

File: doc/sphinx/practical-tools/utilities.rst (hunk 6 above)
Comment (herbelin) [reply doc/sphinx/practical-tools/utilities.rst:459]: Could this sentence be embedded in a previous sentence, as in:
```
the advantage of a wrapper (compared e.g. to including the generated makefile with an include directive) is that it provides a way independent of the version of Coq to regenerate a ``Makefile``
```
?

Comment (JasonGross) [reply doc/sphinx/practical-tools/utilities.rst:459]: One drawback is that you can't have your own target for generating the included Makefile, because the generated Makefile includes a target regenerating itself, and there's not a good way to override that target.

Comment (herbelin) [reply doc/sphinx/practical-tools/utilities.rst:459]: So:
```
... the advantage of a wrapper (compared to directly calling the generated makefile
or to including the generated makefile with an include directive) is that it provides a way 
independent of the version of Coq to regenerate a ``Makefile`` ...
```
?

File: doc/sphinx/practical-tools/utilities.rst (hunk 5 above)
Comment (gares) [reply doc/sphinx/practical-tools/utilities.rst:142]: OK, let me rephrase: "Makefile.local is mainly to extend existing targets or set variables. You can also add new targets".
The current text does not even mention what I consider to be the main use cases.

File: doc/sphinx/practical-tools/utilities.rst (hunk 6 above)
Comment (gares) [reply doc/sphinx/practical-tools/utilities.rst:459]: Saying that A has advantages over B is not the same thing as saying that you should no do B.
The reasons are other, like that `inclusion` is not modular, all variables and rules are injected into the same name space.
This is why there is Makefile.conf that only include variables assignment and all variables are named COQMF_*, that one you can include.

**Comment by Zimmi48** [comment]
@gares Can you be the assignee for this PR?

File: doc/sphinx/practical-tools/utilities.rst (hunk 5 above)
Comment (herbelin) [reply doc/sphinx/practical-tools/utilities.rst]: @gares (and @JasonGross): I pushed an update but I'm not fully sure I expressed correctly what you intended. Don't hesitate to make a suggestion.

**Comment by gares** [comment]
@herbelin I'm not happy with this PR. It essentially moves text around and performs changes that are a little controversial.
IMO #12903 is not a bug, is a question, and the answer is "Because you may already have a file called `Makefile` as suggested in this section of the manual".

I propose this PR and close the bug report with the message above.

**Comment by herbelin** [comment]
One of my points is that I met beginners who did not understand why a wrapper was needed, and did not know either about the `-f` option of `make` either, so they were stuck. This is why I added a documentation of `-f` in #12884 but then I realized that the intended use was not to run `make -f` but rather to use a wrapper.

I don't care about the exact text, but I still believe that the recommendation of using a wrapper should be made clearer. Maybe simply replacing the sentence "Run ``make`` with the ``-f CoqMakefile`` option to use ``CoqMakefile``" with a link to the section "Reusing/extending the generated Makefile" would be enough (replacing there "Reusing" with "Using"?) What do you think?

File: doc/sphinx/practical-tools/utilities.rst [hunk 7]
```diff
@@ -92,14 +92,51 @@ CoqMakefile
...
+
+      # This should be the last rule, to handle any targets not declared above
+      %: invoke-coqmakefile
+              @true
+
+The advantage of a wrapper, compared to directly calling the generated
+``Makefile`` or to including it with an include directive, is that it
+provides a target independent of the version of Coq to regenerate a
+``Makefile`` specific to the current version of Coq. Additionally, the
+master ``Makefile`` can be extended with targets not specific to Coq.
```
Comment (gares) [inline doc/sphinx/practical-tools/utilities.rst:139]:
```suggestion
The advantage of a wrapper, compared to directly calling the generated
``Makefile``, is that it
provides a target independent of the version of Coq to regenerate a
``Makefile`` specific to the current version of Coq. Additionally, the
master ``Makefile`` can be extended with targets not specific to Coq.
Including the generated makefile with an include directive is
discouraged, snice the contents of this file, including variable names and
status of rules, may change in the future.
```

Comment (herbelin) [reply doc/sphinx/practical-tools/utilities.rst:139]: For me this is ideal. Let me already commit the suggestion.

**Review by gares — APPROVED** [review]

**Comment by gares** [comment]
@coqbot merge now
